README
TM-Corpus
=========

This package is additional to the TM (Topic Maps) suite on CPAN.  It
enriches a topic map instance with a document corpus. Content can be
retrieved from the topic map and/or can be harvested from the Internet
(for external references). Such a combo (map + document) can exist
also in persistent form, just to avoid the constant downloading of
resources.

As example application I have packaged a Plucene based fulltext query
trait. If you attach it to a corpus, it becomes queryable.


Copyright (C) 2008 by Robert Barta

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

Supported by a research grant ARCS 2008, http://www.arcs.ac.at/



Changes
Revision history for Perl extension TM::Corpus.

	~ signals deprecation
	! signals compatibility break

0.11  Tue May  5 10:44:36 CEST 2009
      - fix: when using MLDBM version, harvest would not memorize

0.10  Tue Apr 28 11:06:42 CEST 2009
         - fix: Workbench still makes illegal use of UserAgent component

0.09  Sun Apr 26 17:05:26 CEST 2009
         - updated makefile

0.08  Sun Apr 26 15:47:27 CEST 2009
         - removed UA from object (clashes with Storable which complains about not being able to store CODE)

0.07  Sun Nov  2 04:43:16 CET 2008
	- moved Workbench closer to TS (e.g. internet)
	- rearranged tmpfile handling (I hate files) for bigger compatibility out there
	- reduced version requirements for better Debianess

0.06  Sat Nov  1 10:04:13 CET 2008
	- add missing dependency

0.05  Fri Oct 31 20:21:04 CET 2008
	- moved MLDBM version of Corpus to MLDBM::Sync (otherwise clash with TM::Materialized::MLDBM)
	- added directory method to Plucene to control where the index is
	! index of Plucene package does not take the directory as parameter, there is now a dedicated 'directory' method
	- MLDBM is now expecting no map as parameter if the DBM file already exists
	- Workbench commands moved a bit toward TempleScript

0.04  Wed May 28 16:42:13 CEST 2008
	- add: TM::Corpus::Document
	- add: tokenizing and feature detection in Document
	- add: functionality in Corpus to add/retract documents
	- modified: now corpus also include subject address and subject indicators as resources

0.03  Sun May 18 14:22:38 CEST 2008
	- TM::Corpus::SearchAble::Plucene: doc fixes
	- added: TM::Corpus::MLDBM as persistent implemention
	- added: TM::Workbench::Plugin::Corpus as plugin for the tm workbench
	- modified: uses $TM::log object
	- added: decent README file

0.02  ????????????????????????
	- intermediate test version

0.01  Sun Apr 13 11:46:22 2008
	- original version; created by h2xs 1.23 with options
		-X -n TM::SearchAble -b 5.8.0



t/lib/LWP/Mock.pm
package LWP::Mock;

sub new {
    return bless {}, shift;
}

sub get {
    my $self = shift;

    use HTTP::Response;
    use HTTP::Status;
    my $resp = HTTP::Response->new( HTTP::Status->RC_OK, "All well and good", undef, '<html />' );
    $resp->header ('Content-Type' => 'text/html' );
    return $resp;
}

1;


Makefile.PL
use 5.008;
use ExtUtils::MakeMaker;

WriteMakefile(
	      NAME          => 'TM::Corpus',
	      VERSION_FROM  => 'lib/TM/Corpus.pm', # finds $VERSION
	      PREREQ_PM     => {
		  'TM'          	=> 1.45,
		  "Test::More"  	=> 0.47,
		  "Test::Pod"   	=> 1.00,
		  'LWP'         	=> 2.033,
		  'Plucene'     	=> 1.24,
		  'File::Path'  	=> 1.08,
		  'File::Temp'  	=> 0.16,
		  'BerkeleyDB'  	=> 0.26,
		  "MLDBM"       	=> 2.01,
		  'Class::Accessor'     => 0.30,
		  'HTML::Entities'      => 1.35,
		  'Test::Pod::Coverage' => 1.08,
		  'Digest::MD5'         => 2.36,
		  'Regexp::Common::Email::Address' => 1.01,
		  'Regexp::Common'      => 2.120,
	      },
	      ABSTRACT_FROM  => 'lib/TM/Corpus.pm', # retrieve abstract from module
	      AUTHOR         => 'Robert Barta <drrho@cpan.org>');

package MY;

sub depend {
    return <<MAKE

dependencies:
	scandeps.pl `grep .pm MANIFEST` `grep bin/ MANIFEST`| grep -v undef

manifesto:
	find lib/ -iname '*.pm' | sort
	find t    -iname '*.t'  | sort

#$(DISTVNAME).tar$(SUFFIX)

upload:
	cpan-upload -verbose TM-Corpus-\$(VERSION).tar.gz

inject:
	mcpani --add --module TM::Corpus --authorid DRRHO --modversion \$(VERSION) --file TM-Corpus-\$(VERSION).tar.gz
	sudo mcpani --inject
	echo "reload index" | cpan

debian:
	(cd ; cd packages; rm -rf TM-Corpus* ; dh-make-perl --cpan-mirror file:/usr/local/share/minicpan --build --cpan TM::Corpus --notest --core-ok)

MAKE
}



lib/TM/Corpus.pm
package TM::Corpus;

use strict;
use warnings;
use Data::Dumper;

require Exporter;
use base qw(Exporter);

use TM::Literal;
use TM::Corpus::Document;

=pod

=head1 NAME

TM::Corpus - Topic Maps, Document Corpus

=head1 SYNOPSIS

   use TM;
   my $tm = ...

   use TM::Corpus;
   my $co = new TM::Corpus (map => $tm);    # bind with map

   $co->update;                             # copy all content from the map
   $co->harvest (new LWP::UserAgent);       # add documents from the Internet

=head1 ABSTRACT

This package connects a topic map instance and a document corpus into one container.

=head1 DESCRIPTION

A I<corpus> is normally a set of documents. A I<topic map based corpus> is a set of documents,
internal or external to a topic map.

Whenever your topic map is stable, you can first update the corpus with the content and then let an
user agent download all documents which are mentioned in the map. With this data corpus you can then
do any number of things, one of them having it fulltext-searched.

=head1 INTERFACE

=head2 Constructor

The constructor accepts a hash as parameter with the following keys:

=over

=item C<map> (mandatory)

The value must be a L<TM> object. Any map should do.

=item C<ua> (optional)

You can pass in your own L<LWP::UserAgent> object. That is used when you ask to C<harvest> the
documents behind occurrence URLs. If you omit that, a stock object will be generated.

=back

=cut

sub _init_ua {
    use LWP::UserAgent;

    my $ua = LWP::UserAgent->new;
    $ua->agent ('TM Harvester 0.1');
    $ua->timeout(5);
    $ua->env_proxy;
    return $ua;
}

sub new {
    my $class = shift;
    my %options = @_;

    $options{map} or $TM::log->logdie (__PACKAGE__ . ": you have to provide a map to attach to");
    $options{resources} = {};
    $options{deficit}   = {};
    return bless \%options, $class;
}

=pod

=head2 Methods

=over

=item B<map>

my I<$tm> = I<$co>->map

Read-only access to the underlying map.

=cut

sub map {
    my $self = shift;
    return $self->{map};
}

=pod

=item B<resources>

Read-only access to the data. Probably not wise to use, but here it is.

=cut

sub resources {
    my $self = shift;
    return $self->{resources};
}

=pod

=item B<update>

I<$co> = I<$co>->update

I<$co>->update

This method synchronizes all I<data> from the map into the corpus. The underlying map is the
authoritative source, but when it is modified, the corpus is NOT automatically updated. Instead, you
should invoke this method at a suitable time.

=cut

sub update {
    my $self = shift;
    my $tm   = $self->{map};
    my $res  = $self->{resources};

    #-- collect things from the map
    # 1) assertion values
    foreach my $a (grep { $_->[TM->KIND] != TM->ASSOC } $tm->match_forall ()) {   # go for all occs and names
	next if $res->{ $a->[TM->LID] };                                          # we already have a value for that one

	my ($tid, $val) = @{ $a->[TM->PLAYERS] };                                 # get the topic id and the whole value
	if ($val->[1] eq TM::Literal->URI) {                                      # for references we only store these
	    $res->{ $a->[TM->LID] } = new TM::Corpus::Document ({
		                        aid => $a->[TM->LID],
					ref => $val->[0],
					tid => $tid,
					attachment => 'characteristics',
				        });
	} else {                                                                  # for values the value itself
	    $res->{ $a->[TM->LID] } = new TM::Corpus::Document ({
		                        aid  => $a->[TM->LID],
					val  => $val->[0],
					mime => 'text/plain',   # TODO: where to get a decent MIME from?
					tid  => $tid,
					attachment => 'characteristics',
				        });
	}
    }
    # 2) topics with subject locators, indicators
    foreach my $t ($tm->toplets (\ '+all -infrastructure')) {
	my $tid = $t->[TM->LID];

	_inject ($res, new TM::Corpus::Document ({ref => $t->[TM->ADDRESS ], tid => $tid, attachment => 'address', }))
	    if $t->[TM->ADDRESS];
	sub _inject {
	    my $res = shift;
	    my $doc = shift;
	    $res->{ $doc->ref } = $doc; 
	}
	map { _inject ($res, new TM::Corpus::Document ({ref => $_,           tid => $tid, attachment => 'indication', }) ) }
	    @{ $t->[TM->INDICATORS] };
    }

    # cleanup leftover of values from assertions which do not exist anymore
    map  { delete $res->{$_} }                                                    # delete those
    grep { ! ( $tm->retrieve ($_) || $tm->tids ($_) || $tm->tids (\ $_)) }        # which do not exist as assertions, addrs or indics
    keys %$res;                                                                   # in our resources
    $self->{resources} = $res;                                                    # this makes MLDBM happy
    return $self;
}

=pod

=item B<harvest>

I<$co> = I<$co>->harvest

I<$co>->harvest

I<$co>->harvest (I<$ua>)

This method uses the defined user agent to resolve all URLs within the underlying map and to load
the content locally. All network related modalities (timeout, limits, etc.) have to be implemented
via the user agent.

=cut

sub harvest {
    my $self = shift;
    my $ua   = shift || _init_ua;

    my $res  = $self->{resources};
    foreach my $r (keys %$res) {
	my $rr = $res->{$r};
	next if     $rr->val;
	next unless $rr->ref;
	my $resp;
#	eval {
	    $resp = $ua->get ( $rr->ref );
#	}; die $@ if $@;                                             # propagate exception for now

	if ($resp->is_success) {
	    $rr->val  ( $resp->content );
	    $rr->mime ( $resp->header ('Content-Type') );
	} else {
	    $rr->{_fetch_fails}++;
	    $self->{deficit}->{$r} = { url => $rr->ref, fails => $rr->{_fetch_fails}++ };
	}
	$rr->{_last_attempt} = time;
	$rr->{_last_code}    = $resp->code;
	$res->{$r} = $rr;                                            # this is important if we TIE this!
    }
    $self->{resources} = $res;                                       # this is important if we TIE this!
    return $self;
}

=pod

=item B<deficit>

I<$deficit> = I<$co>->deficit

This method returns all the URL references which could not be resolved successfully during all
previous invocations of C<harvest>. It is a hash (reference) with the assertion ID as key and a
C<url>, C<fails> combo as value.

Example:

    warn "damn" if keys %{ $co->deficit };

=cut

sub deficit {
    my $self = shift;
    return $self->{deficit};
}

=pod

=item B<inject>

I<$co>->inject (I<$tid> => I<$doc>, ...)

This method injects documents into the corpus. For each of these you have to provide a topic
identifier (tid, see L<TM>) and a L<TM::Corpus::Document> instance.

=cut

sub inject {
    my $self = shift;
    my $tm   = $self->{map};
    my $res  = $self->{resources};

    while (@_) {                                             # they always come in pairs
	my $tid = shift;
	my $doc = shift;

	$tid = $tm->internalize ($tid);                      # really get a working topic id
	use Digest::MD5 qw(md5);
	my $uri = $doc->ref || 'inline:'.md5($doc->val);
	$tm->internalize ($tid => $uri);                     # add a subject locator if doc had a URL
	$res->{ $uri } = $doc;                               # attach it to the resources
	$doc->{ tid }  = $tid;                               # have it there whether you like it or not
    }
}

=pod

=item B<extract>

I<@docs> = I<$co>->extract (I<$tid>, ...)

Given a list of topic identifiers, the subject addresses of these will be use to identify the
underlying documents. These are then returned in a list.

=cut

sub extract {
    my $self = shift;
    my $tm   = $self->{map};
    my $res  = $self->{resources};

    return  map  { $res->{$_} }
            grep { defined $_ }
	    map  { $_->[TM->ADDRESS] }
	    grep { defined $_ }
            map  { $tm->toplet ( $_ ) }
            $tm->tids (@_)
            ;
}

=pod

=item B<eject>

<$co>->eject (I<$tid>, ...)

This method removes all documents which are subjects of the topics handed in.  The topics are
removed as well from the underlying topic map.

=cut

sub eject {
    my $self = shift;
    my $tm   = $self->{map};
    my $res  = $self->{resources};

    map  { delete $res->{$_} }
       grep { defined $_ }
       map  { $_->[TM->ADDRESS] }
       grep { defined $_ }
       map  { $tm->toplet ( $_ ) }
       $tm->tids (@_);
    map { $tm->externalize ($_) }
       $tm->tids (@_);
}

=pod

=item B<features>

I<$fs>           = <$co>->features (I<%options>)

(I<$fs>, I<$vs>) = <$co>->features (I<%options>)

@@@@ total @@@

This method computes a hash (reference) of feature values inside the corpus. Optionally the method
also returns a list of all feature vectors from which the feature set has been computed.

The extracting and tokenizing options are those in L<TM::Corpus::Document>. Additionally you can
define cut-off points to remove the most frequent and the least frequent features.

=over

=item C<low> (default: 0.01)

=item C<high> (default: 1.0)

=back

@@ option nr @@@@

0.01, 1.00 constants

=cut

sub features {
    my $self = shift;
    my %options = @_;

    my @fvs = map  { $_->features (%options) }                                  # collect individual feature vectors
              grep { $_->val }
              map  { $self->{resources}->{$_} }
              keys %{ $self->{resources} };                                     # strange: values would not work....

    # TODO: if this is a name, treat it differently, than a value occurrence, than a ref
    # TODO: 20:20:60

    my %features;                                                               # consolidate all features
    foreach my $fv (@fvs) {
	map { $features{$_} += $fv->{$_} } keys %{ $fv };
    }
#warn Dumper \%features;
    #-- normalize ---------------------------------------------------------------------
    my $total;
    map { $total += $features{$_} } keys %features;
    map { $features{$_} /= $total } keys %features;
#warn Dumper \%features;

    my $low  = defined $options{low}  ? $options{low}  : 0.01;                  # everything below this will be ignored
    my $high = defined $options{high} ? $options{high} : 1.00;                  # everything above this will be ignored
    #-- cutoff features too frequent or too unfrequent --------------------------------
    map { delete $features{$_} if $features{$_} < $low  } keys %features;
    map { delete $features{$_} if $features{$_} > $high } keys %features;

    if (defined $options{nr}) {                                                 # absolute length cutoff
	my @labels = sort { $features{$b} <=> $features{$a} }                   # sort by feature strength
	             keys %features;                                            # find all the values for these features (by popularity desc)
	@labels = @labels[0..$options{nr}-1] if $options{nr} <= scalar @labels; # take a slice, if there is more
	my %f2  = map { $features{$_} ? ($_ => $features{$_}) : ()} @labels;    # copy those which should count
	%features = %f2;
    }
    return (\%features, \@fvs, $total);
}

=pod

=back

=head1 SEE ALSO

L<TM::Corpus::MLDBM>, L<TM::Corpus::SearchAble>

=head1 COPYRIGHT AND LICENSE

Copyright 200[89] by Robert Barta, E<lt>drrho@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify it under the same terms as Perl
itself.

=cut

our $VERSION = '0.11';

1;

__END__


lib/TM/Corpus/MLDBM.pm
package TM::Corpus::MLDBM;

use strict;
use warnings;

use base 'TM::Corpus';

use Data::Dumper;

use BerkeleyDB ;
use MLDBM qw(MLDBM::Sync::SDBM_File Storable) ;
use MLDBM::Sync;
use Fcntl qw(:DEFAULT);

=pod

=head1 NAME

TM::Corpus::MLDBM - Topic Maps, MLDBM backed document corpus

=head1 SYNOPSIS

    # see TM::Corpus
    # it is the same interface & procedure, you only
    # use TM::Corpus::MLDBM instead of TM::Corpus

=head1 ABSTRACT

This package connects a topic map instance and a document corpus into one container. All document
portions (not the topic map itself) will be made persistent in a MLDBM database.

=head1 DESCRIPTION

This subclass of L<TM::Corpus> exposes the same functionality as its mother class. The
only difference is that all document aspects (not the topic map underlying) are stored
persistently.

Obviously when you harvest documents off the Internet this will take significant resources,
especially when the map is rich and points to many external documents. In order to safe resources,
you can use this class as an in-place replacement to L<TM::Corpus>. As a consequence all harvested
documents will remain until you remove them explicitely.

B<NOTE>: Only when the object goes out of scope (actually the destructor is invoked), then you can
be certain that all content has been sync'ed out to disk.

=head1 INTERFACE

=head2 Constructor

The constructor accepts a hash as parameter with the following keys:

=over

=item C<file> (default: none)

Name of a (.dbm) file in the local file system where the MLDBM is stored. If it does not exist, it
will be silently created.

=back

=cut

sub new {
    my $class    = shift;
    my %options  = @_;

    my $file     = delete $options{file} or $TM::log->logdie ("no file specified");
    my %self;
    if (-e $file || -e $file . '.pag') {                                            # file does exist already
#warn "file exists";
        tie %self, 'MLDBM::Sync', $file, O_RDWR, 0600                               # tie the whole thing
            or $TM::log->logdie ( "Cannot tie to DBM file '$file: $!");
                                                                                    # oh, we are done now
#	$self{ua} = $options{ua} || TM::Corpus::_init_ua;
    } else {                                                                        # no file yet
#warn "file not exists $file!";
        tie %self, 'MLDBM::Sync', $file, O_RDWR|O_CREAT, 0600                       # tie now
	     or $TM::log->logdie ( "Cannot create DBM file '$file: $!");
	%self = %{ $class->SUPER::new (%options) };                                 # create an empty beast
    }
    return bless \%self, $class;
}

sub DESTROY {                                                                       # if an object went out of scope
    my $self = shift;
    untie %$self;                                                                   # release the tie with the underlying resource
}

=pod

=head1 SEE ALSO

L<TM::Corpus>, L<TM::Corpus::SearchAble>

=head1 COPYRIGHT AND LICENSE

Copyright 200[89] by Robert Barta, E<lt>drrho@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify it under the same terms as Perl
itself.

=cut

our $VERSION = '0.04';

1;

__END__


}





lib/TM/Corpus/SearchAble.pm
package TM::Corpus::SearchAble;

use strict;
use warnings;
use Data::Dumper;

use Class::Trait 'base';

=pod

=head1 NAME

TM::Corpus::SearchAble - Topic Maps, Abstract Trait for searching

=head1 SYNOPSIS

   use TM;
   my $tm = ....                               # get map from somewhere

   use TM::Corpus;                             # see this package
   my $co = new TM::Corpus (map => $tm)        # bind map with document repository
            ->update                           # mandatory
            ->harvest;                         # optional

                                               # attach searchable behaviour
   Class::Trait->apply ($co => 'TM::Corpus::SearchAble::SomeImplementation');
   $co->directory ('/where/store/index/');
   $co->index;                                 # build index

   warn Dumper $co->search ('content:"BBB"');  # search for something

   # if you already have an indexed corpus, then
   my $co = ....
   Class::Trait->apply ($co => 'TM::Corpus::SearchAble::SomeImplementation');
   $co->directory ('/where/index/is/stored/');

   warn Dumper $co->search ('content:"BBB"');  # search for something


=head1 DESCRIPTION

This package is only abstract and it defines the minimal interface a
trait which provides search functionality has to honor.

=head1 INTERFACE

=over

=item B<index>

I<$co>->index (...)

This method creates an index. Individual implementations may need additional parameters. The method
returns the object itself.

=item B<search>

I<@results> = @{ I<$co>->search (I<$phrase>) }

This method takes a search phrase as input and delivers a list (reference) of results. Individual
implementations will have special syntaxes for the search phrase, but they have to honor the
following fields:

=over

=item C<content> (tokenized): the content of the referred documents, any names or the values of
occurrences (if they are not URIs)

=item C<ref> (tokenized): the URI references in occurrences

=item C<tid> (as-is): the toplet identifier where the value is attached

=item C<aid> (as-is): the assertion identifier where the value is part of

=back

Each result entry is a list (reference) containing (in sequence):

=over

=item C<aid>: assertion identifier

=item C<tid>: local toplet identifier (to which that assertion belongs)

=item C<baseuri>: the baseuri of the map

=back

=back

=head1 COPYRIGHT AND LICENSE

Copyright 200[8] by Robert Barta, E<lt>drrho@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify it under the same terms as Perl
itself.

=cut

our $VERSION = '0.03';

1;

__END__


lib/TM/Corpus/Document.pm
package TM::Corpus::Document;

use strict;
use warnings;

use Data::Dumper;

use base qw(Class::Accessor);
TM::Corpus::Document->mk_accessors(qw(tid aid ref val mime _fetch_fails _last_attempt));

=pod

=head1 NAME

TM::Corpus::Document - Topic Maps, Document

=head1 SYNOPSIS

   use TM::Corpus::Document;
   my $d = new TM::Corpus::Document ({ mime => 'text/plain',
                                       val  => 'this is some text' });
   # accessors
   $val  = $d->val ('new text');
   $mime = $d->mime ('new/mime');
   $url  = $d->ref ('http://somewhere/some.txt');

   my @tokens = $d->tokenize; # leaving defaults

   # using some predefined tokenizing steps, in this order
   my @tokens = $d->tokenize (tokenizers => 'NUMBER QUOTER COM&BO');

   # using negative ones (i.e. throw things away)
   my @tokens = $d->tokenize (tokenizers => 'COM&BO COM-BO -INTERPUNCT');

   # using filters (detect numbers and throw them away
   my @tokens = $d->tokenize (tokenizers => 'NUMBER !NUMBER');

   # get also debugging output
   my @tokens = $d->tokenize (tokenizers => 'NUMBER TAP !NUMBER TAP');

   # define your own filters
   $TM::Corpus::Document::FILTERS{'!4LETTER'} = 
                sub { $_ = shift; return length($_) == 4 ? '' : $_; };
   my @tokens = $d->tokenize (tokenizers => 'WORDER !4LETTER');

   # collect features, here single tokens and two subsequent tokens
   my %features = $d->features (tokenizers  => '...', 
                                featurizers => 'TOKEN1 TOKEN2')

=head1 ABSTRACT

This package implements I<documents>, i.e. document pertinent information, such as its content, the
corresponding MIME type, maybe a reference to the document if it has one.

Most notable is functionality to find the I<tokens> (i.e. word substrings) and derive from these
also a feature vector for the document.

=head1 DESCRIPTION

=head1 INTERFACE

=head2 Constructor

The constructor expects a hash B<reference> with one or more of the following fields:

=over

=item C<ref>

A URI string to refer to the network address of the document. In Topic Maps parlor this
will be the I<subject locator> for the document topic.

=item C<val>

The character stream associated with the document.

=item C<mime>

The MIME type of the content.

=back

=head2 Methods

=over

=item B<ref>

Accessor for the C<ref> component of the document. Nothing happens with the other components.

=item B<val>

Accessor for the C<val> component of the document. Nothing happens with the other components.

=item B<mime>

Accessor for the C<mime> component of the document. Nothing happens with the other components.

=item B<tokenize>

This method returns a list reference to recognized tokens.

To generate this, the method will first find an I<extractor> according to the document's MIME
type. That will extract text, but also relevant meta data, such as title, length, etc. Some
extractors are predefined; you can get a list with

   perl -MTM::Corpus::Document -e 'warn join ",", keys %TM::Corpus::Document::EXTRACTORS;'

The extractor can also be overridden:

   $d->tokenize (extractor => sub { ... });

It gets the value (content) as first parameter.

In a second step the content stream of the document is analyzed for patterns, such as numbers, dates
or words. To control from the outside what is relevant and what should be done in which order, this
is specified with a simple language.

Example:

   $d->tokenize (tokenizers => 'COM&BO COM-BO');

I<Positive> tokenizers detect patters and bless them as valid tokens which will not be further
analyzed or questioned:

=over

=item C<WORDER>: detects word in current locale

=item C<QUOTER>: detects substrings wrapped in ""

=item C<NUMBER>: detects decimal numbers

=item C<DATE>: detects date specification in current locale (NOT IMPLEMENTED!)

=item C<COM&BO>: detects patterns like AT&T

=item C<COM-BO>: detects patterns like T-Mobile

=item C<Capitalize>: detects capitalized words

=back

Negative tokenizers detect patterns and immediately throw them away:

=over

=item C<-WORDER>: everything which is left as text fragment is suppressed

=item C<-QUOTER>: quoted text is suppressed

=item C<-NUMBER>: decimal numbers are suppressed

=item C<-INTERPUNCT>: interpunctations characters are suppressed

=back

I<Filters> take existing tokens and either modify then, suppress them or pass them through (and
suppress everything else).

=over

=item C<!NUMBER>: number tokens are replaced with empty tokens

=back

You can override and extend tokenizers and filters by tampering with the hashes C<%TOKENIZERS> and
C<%FILTERS>. You can hook in, for instance a stopword list like this:

    my %stops =  map { $_ => 1 } qw(Terror CIA HLS);
    $TM::Corpus::Document::FILTERS{'!STOPS'} = 
                 sub { $_ = shift; return $stops{$_} ? '' : $_; };

    $d->tokenize (tokenizers => ' .... !STOPS ....');

=cut

our %EXTRACTORS = (
		   'text/plain' => sub {
		       return { content => $_[0],
				size    => length $_[0],
				mime    => 'text/plain', }
		   },
		   'text/html'  => sub { 
		       $_ = shift;
		       s/<script.*?>.*?<\/script>//isg;  # fu**ing JS
		       s/<!--.*?-->//sg;                 # no comments please
		       s/<.*?>/ /sg;                     # detag
		       s/\r/\n/g;                        # EOL sanitize
		       s/\n+/\n/g;                       # EOL canonicalize
		       s/\s+/ /g;                        # minimalize blanks

		       use HTML::Entities;
		       decode_entities( $_ );
		       return { content => $_,
				size    => length $_,
				mime    => 'text/html' }
		   }
		   );

# excel .xls => system xlhtml
# word .doc => antiword
# html .html => lynx dump
# PDF  .pdf => pdftohtml -htmlmeta
# bauernpoint .ppt => odt2text 

use Regexp::Common qw /URI/;
use Regexp::Common::Email::Address qw /URI/;
our %PATTERNS = (
		 'QUOTE'       => qr/[\'"](.*?)[\'"]/,
		 'NUMBER'      => qr/(?<!\w)(\d+(\.\d*)?)(?!\w)/,
		 'COM&BO'      => qr/(\w+&\w+)/,
		 'COM-BO'      => qr/(\w+-\w+)/,
		 'CO.M.BO.'    => qr/\b(\w\.\w(\.\w)?\.)/,
		 'INTERPUNCT'  => qr/([\.\!\&\-\|:;,\"\'\?\(\)\[\]\{\}\/\+\-\*])/,
		 'Capitalized' => qr/\b([A-Z]\w+?)\b/,
		 'URI'         => qr/($RE{URI}{-keep})/,
		 'EMAIL'       => qr/($RE{Email}{Address}{-keep})/,

		 'lowercase'   => sub { return lc( $_[0] ); },
		 'UPPERCASE'   => sub { return uc( $_[0] ); },
		 );

sub tokenize {
    my $self    = shift;
    my %options = @_;
    $options{extractor}  ||= $EXTRACTORS{ $self->{mime} };
    die "no extractor" unless $options{extractor};
    $options{tokenizers} ||= '+QUOTE -NUMBER +WORD';
    
sub _tok {
    my $p = shift;
    if (ref $p eq 'CODE') { 
	return &$p (shift); 
    } else {
	$_ = shift; s/$p/<$1>/g; return $_;
    }
}

sub _filt {
    my $p = shift;
    if (ref $p eq 'CODE') { 
	return &$p (shift); 
    } else {
	$_ = shift; return $_ if /$p/;
    }
}

sub _notok {
    my $p = shift;
    if (ref $p eq 'CODE') { 
	return &$p (shift);
    } else {
	$_ = shift; s/$p//g; return $_;
    }
}

sub _word {
    use locale;
    return join "", map { "<$_>" } grep { length($_) } split /\s*\b\s*/, $_[0];
}

    my $d = &{ $options{extractor} } ($self->{val});
    my $c = '>' . $d->{content} . '<';              # maybe do the tokenizing on all other components later?
    $c =~ s/\s+/ /sg;                               # canonicalize blanks (and newlines?)
    foreach my $e (split /\s+/ , $options{tokenizers} ) {
#	warn $e;
	my $p;
	if ($e =~ /^\+WORD/) {
	    $c =~ s/>([^<]+?)</'>'.&_word($1).'<'/egs;

	} elsif ($e eq '|') {
	    warn $c;

	} elsif (($e =~ /^\+(.+)/)    && ($p = $PATTERNS{$1})) {
	    $c =~ s/>([^<]+?)</'>'.&_tok($p, $1).'<'/egs;

	} elsif (($e =~ /^\-(.+)/)    && ($p = $PATTERNS{$1})) {
	    $c =~ s/>([^<]+?)</'>'.&_notok($p, $1).'<'/egs;

	} elsif (($e =~ /^\-<(.+)>/)  && ($p = $PATTERNS{$1})) {
	    $c =~ s/<(.*?)>/'<'.&_notok($p, $1).'>'/egs;

	} elsif (($e =~ /^\+<(.+)>/)  && ($p = $PATTERNS{$1})) {
	    $c =~ s/<(.*?)>/'<'.&_filt($p, $1).'>'/egs;

	} else {
	    die "unknown operator '$e': ".Dumper \%PATTERNS;
	}

    }
    return [ grep { $_ =~ /\S/ } ($c =~ /<([^<]+?)>/g) ];
}

=pod

=item C<features>

This method computes the feature vector from a document. It accepts all parameters from method
B<tokenize> as it will invoke this first. Additionally you can specify I<how> to tokenize

  my %fv = $d->features (tokenizers  => 'QUOTER NUMBER WORDER', 
                         featurizers => 'TOKEN1 TOKEN2');

Following tokenizers are defined:

=over

=item C<TOKEN1>: occurrences of single tokens are counted in the document

=item C<TOKEN2>: occurrences of two subsequent tokens in the document are counted

=item C<TOKEN3>: group of 3 are counted

=item C<MIME>: the MIME type is converted into some numeric value

=back

You can extend or modify the C<%FEATURIZERS> hash to add your own featuritis.

=cut

our %FEATURIZERS = (
		    'TOKEN1' => sub { 
			my $d = shift; my $l = $d->{_tokens};
			my %F;
			for (my $i = 0; $i <= $#$l; $i++) {
			    $F{ $l->[$i] }++;
			}
			return %F;
		    },
		    'TOKEN2' => sub { 
			my $d = shift; my $l = $d->{_tokens};
			my %F;
			for (my $i = 0; $i <= $#$l - 1; $i++) {
			    $F{ $l->[$i] .' '. $l->[$i+1] }++;
			}
			return %F;
		    },
		    'TOKEN3' => sub { 
			my $d = shift; my $l = $d->{_tokens};
			my %F;
			for (my $i = 0; $i <= $#$l - 2; $i++) {
			    $F{ $l->[$i] .' '. $l->[$i+1] .' '. $l->[$i+2]}++;
			}
			return %F;
		    },
		    'MIME' => sub {
			$_ = shift->{mime};
			m|text/plain| && return (mime => 1);
			m|text/html|  && return (mime => 2);
		    }
		    );

sub features {
    my $self    = shift;
    my %options = @_;
    $options{featurizers} ||= qw(TOKEN1);

    $self->{_tokens} = $self->tokenize (%options);
#    warn Dumper $self;

    my %fs = (); # will contain all features
    foreach my $f (split /\s+/ , $options{featurizers}) {
	%fs = (%fs, &{ $FEATURIZERS{ $f }} ($self));
    }
#    warn "feature ".Dumper \%fs;
    return \%fs;
}

=pod

=back

=head1 NOTES

No. Plucene tokenizing was NOT helpful.

=head1 SEE ALSO

L<TM::Corpus>

=head1 COPYRIGHT AND LICENSE

Copyright 200[8] by Robert Barta, E<lt>drrho@cpan.orgE<gt>

This library is free software; you can redistribute it and/or modify it under the same terms as Perl
itself.

=cut

our $VERSION = '0.02';

1;

__END__

                                   
our %TOKENIZERS = (
		   'WORDER' 	 => sub { use locale;
					  return join "", map { "<$_>" } grep { length($_) } split /\s*\b\s*/, shift },
		  '-WORDER' 	 => sub { $_ = shift; s/(.*?)//gs; return $_ },
		   'QUOTER' 	 => sub { $_ = shift; s/"(.*?)"/<$1>/gs; return $_ },
		  '-QUOTER' 	 => sub { $_ = shift; s/".*?"//gs; return $_ },
		   'NUMBER' 	 => sub { $_ = shift; s/(?<!\w)(\d+(\.\d*)?)(?!\w)/<$1>/g; return $_ },
		  '-NUMBER' 	 => sub { $_ = shift; s/(?<!\w)\d+(\.\d*)?(?!\w)//g; return $_ },
		   'COM&BO' 	 => sub { $_ = shift; s/(\w+&\w+)/<$1>/g; return $_; },
		   'COM-BO' 	 => sub { $_ = shift; s/(\w+-\w+)/<$1>/g; return $_; },
		   'CO.M.BO.'    => sub { $_ = shift; s/\b(\w\.\w(\.\w)?\.)/<$1>/g; return $_;},
		   '-INTERPUNCT' => sub { $_ = shift; s/[\.\&\-\|:;,\"\'\?\(\)\[\]\{\}\/\+\-\*]/ /g; return $_; },
		   'Capitalized' => sub { $_ = shift; s/\b([A-Z]\w+?)\b/<$1>/g; return $_; },
		   'URI'         => sub { $_ = shift; 
					  use Regexp::Common qw /URI/;
					  s|$RE{URI}{-keep}|<$1>|g;
					  return $_; },
		   '-URI'        => sub { $_ = shift; 
					  use Regexp::Common qw /URI/;
					  s|$RE{URI}||g;
					  return $_; },
		   'EMAIL'       => sub { $_ = shift; 
					  use Regexp::Common::Email::Address qw /URI/;
					  s|$RE{Email}{Address}{-keep}|<$1>|g;
					  return $_; },
		   '-EMAIL'      => sub { $_ = shift; 
					  use Regexp::Common::Email::Address qw /URI/;
					  s|$RE{Email}{Address}||g;
					  return $_; },
		   );

our %FILTERS = (
		'!NUMBER'    => sub { $_ = shift; s/^\d+(\.\d*)?$//gs; return $_ },
		'!lowercase' => sub { return lc( shift ); },
#		'TAP'     => sub { warn $_[0]; return $_[0]; },
		);

